### PR TITLE
Refactor Downloader and Dependency Manager interfaces

### PR DESCRIPTION
**Pull request by IKukhta**

### CODE REVIEW

File: pkg/depman/depman.go [hunk 1]
```diff
@@ -62,24 +65,24 @@ func NewWithDownloader(bd *bundle.Bundle, dl downloader.Downloader) (DependencyM
...
-		for _, idxDepName := range dm.RootBundle.Index.Depends {
-			depName, _ := bundle.ParseIndexDependency(idxDepName)
-			if _, ok := replaced[depName]; ok {
-				continue
-			}
-			depends = append(depends, idxDepName)
-		}
-		dm.RootBundle.Index.Depends = depends
-	}
+	// // TODO: Possibly needs refactor
```
Comment (yury-palyanitsa) [inline pkg/depman/depman.go:74]: Can be safely removed if no longer used.

File: pkg/depman/download.go [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+	slog.Info("Discovered dependency", slog.String("bundle", sourceName))
+
+	// Pre-download integrity check
+	if err := dm.checkIntegrity(dm.RootBundle.IndexLock, sourceName, upstreamHash); err != nil {
+		return "", nil, fmt.Errorf("check integrity: %w", err)
+	}
+
+	// go-import consists of space-delimited data with:
```
Comment (yury-palyanitsa) [inline pkg/depman/download.go:65]: Move implementation-specific comments

File: pkg/downloader/downloader.go [hunk 3]
```diff
@@ -1,208 +1,6 @@
...
-	}
-	defer resp.Body.Close()
-
-	return io.ReadAll(resp.Body)
-}
-
-func (dl *goLikeDownloader) parseGoQuery(goQuery string) (string, string, string) {
-	parts := strings.Split(goQuery, " ")
-	return parts[0], parts[1], parts[2]
+	Discover(string, string) (Downloader, string, error)
```
Comment (yury-palyanitsa) [inline pkg/downloader/downloader.go:4]: I think `Discover` is implementation-specific. Maybe use generics here if you want stricter interface?

File: cmd/cti/cti.go [hunk 4]
```diff
@@ -281,7 +278,7 @@ func mainFn() int {
 			slog.Error(`                |                   `)
 		}
 		if errors.As(err, &cmdErr) && cmdErr.Inner != nil {
-			slog.Error("Command failed", slogex.ErrorWithTrace(cmdErr.Inner))
+			slog.Error("Command failed", stacktrace.ErrToSlogAttr(cmdErr.Inner))
```
Comment (viatoriche) [inline cmd/cti/cti.go:281]: Please add an option to ensure duplicates in traces: https://github.com/acronis/go-stacktrace/blob/main/traces.go#L27

File: internal/app/commands/validatecmd/command.go [hunk 5]
```diff
@@ -41,7 +41,7 @@ func (c *cmd) Execute(_ context.Context) error {
 	// TODO: Validation for usage of indirect dependencies
 	if errs := bunman.Validate(bd); errs != nil {
 		for i := range errs {
-			slog.Error("Validation error", slogex.ErrorWithTrace(errs[i]))
+			slog.Error("Validation error", stacktrace.ErrToSlogAttr(errs[i]))
```
Comment (viatoriche) [inline internal/app/commands/validatecmd/command.go:44]: And here : https://github.com/acronis/go-cti/pull/25/files#r1776886730

File: pkg/depman/links.go [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+			return nil
+		}
+
+		raw, err := os.ReadFile(file)
+		if err != nil {
+			return fmt.Errorf("read file: %w", err)
+		}
+		// implies that dependencies are correct in the first place
+		content := replaceCaptureGroup(patchDepsRe, string(raw), "../../"+DependencyDirName, 1)
+		content = replaceCaptureGroup(patchRamlxRe, content, "../../"+bundle.RamlxDirName, 1)
```
Comment (yury-palyanitsa) [inline pkg/depman/links.go:54]: Why not just:
```
patchDepsRe.ReplaceAllLiteralString(raw, "../../$1/")
patchRamlxRe.ReplaceAllString(raw, "../../$1/")
```

Comment (IKukhta) [reply pkg/depman/links.go:54]: .dep could be part of other string so regex is more fair

**Pull request merged**